File: openconjectures/erdos624-h4-finite-replay/checkers/verify_h4.rb
X = [0, 1, 2, 3]
values = Hash.new(0)
[[0,1],[2,3]].each { |p| values[p] = 1 }
[[0,2],[1,3]].each { |p| values[p] = 2 }
[[0,3],[1,2]].each { |p| values[p] = 3 }

subsets = (0..4).flat_map { |k| X.combination(k).map(&:to_a) }
cover = lambda do |threshold|
  X.combination(threshold).all? do |target|
    got = subsets.select { |s| (s - target).empty? }.map { |s| values[s] }.uniq
    got.sort == X
  end
end
raise 'm=2 unexpectedly works' if cover.call(2)
raise 'm=3 failed' unless cover.call(3)
raise 'm=4 failed' unless cover.call(4)
puts 'H(4)=3; m0/m1/m2 lower bound; m3 construction passes 5 triples'

File: openconjectures/erdos624-h4-finite-replay/checkers/verify_h4.py
"""Separate finite replay of the source-exact X={0,1,2,3} H(4) slice."""
from __future__ import annotations

from itertools import combinations

X = (0, 1, 2, 3)
SUBSETS = tuple(frozenset(c) for size in range(5) for c in combinations(X, size))
VALUES = {subset: 0 for subset in SUBSETS}
# A 1-factorization of K4: every triple sees all three pair colors.
VALUES.update({
    frozenset((0, 1)): 1,
    frozenset((2, 3)): 1,
    frozenset((0, 2)): 2,
    frozenset((1, 3)): 2,
    frozenset((0, 3)): 3,
    frozenset((1, 2)): 3,
})


def covers(values: dict[frozenset[int], int], threshold: int) -> bool:
    return all(
        {values[source] for source in SUBSETS if source <= frozenset(target)} == set(X)
        for size in range(threshold, len(X) + 1)
        for target in combinations(X, size)
    )


def main() -> None:
    assert not covers(VALUES, 2)
    assert covers(VALUES, 3)
    assert covers(VALUES, 4)
    print("H(4)=3; m0/m1/m2 lower bound; m3 construction passes 5 triples")


if __name__ == "__main__":
    main()
